pkg/cloud/services/network/subnets.go

```diff
@@ -139,11 +128,12 @@ func (s *Service) reconcileSubnets() error {
 				existingSubnet.ID = sub.ID
 			}
 
+			// Make sure tags are up-to-date.
+			subnetTags := sub.Tags
+
```

Comment:
This is the key change that fixes the issue. By storing the subnet tags from the spec to a local variable, we won't overwrite them when using `DeepCopyInto`.

Reply:
Hi @fiunchinho - is this behavior considering managed and unmanaged scenarios?

Checking the documentation about BYO/unmanaged and tags, it [states][1]:
> Note: All the tagging of resources should be the responsibility of the users and are not managed by CAPA controllers.

My understanding is this block will be reached at once when *unmanaged* (`existingSubnets!=nil`), and later interactions in the reconcile loop when *managed* (with subnets created h[ere ][2], preserving subnets tags defined on spec), considering the documentation that *unmanaged* should not create additional tags, than required from k8s, the tags from spec should be skipped. Is it make sense? (+ @richardcase @nrb  would love to hear your thoughts too)

[1]: https://cluster-api-aws.sigs.k8s.io/topics/bring-your-own-aws-infrastructure#tagging-aws-resources
[2]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/c8eb4797b15e4925c07b05782b90add67f59d148/pkg/cloud/services/network/subnets.go#L197-L201

Reply:
hey @mtulio thanks for the comment. If I understand it correctly, with this approach, when reconciling an unmanaged vpc, the tags in the subnet on AWS are applied to the AWSCluster spec, and new tags in the AWSCluster spec are ignored/never applied into AWS. When reconciling a managed VPC, the tags in the AWSCluster spec are applied to the subnet on AWS.

[In the subnets.go file, line 626](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/network/subnets.go#L664), we only take into consideration the tags defined in the `AWSCluster` spec (called `manualTags` in the function) if the vpc is managed.

So I think this approach is not going against current documentation of BYOI/unmanaged infra.

Also, I think finding the subnet in AWS (`existingSubnets!=nil`) doesn't mean it's unmanaged. When CAPA creates the subnets, on following reconciliations it will find the same subnets on AWS, and they are managed by CAPA. A managed VPC is defined as a VPC with an `id` set, and the `sigs.k8s.io/cluster-api-provider-aws/cluster/$clusterName` set.

Reply:
> hey @mtulio thanks for the comment. If I understand it correctly, with this approach, when reconciling an unmanaged vpc, the tags in the subnet on AWS are applied to the AWSCluster spec, and new tags in the AWSCluster spec are ignored/never applied into AWS. [...]

Yep, that's the current behavior, isn't it aligned with the documentation for BYOI/unmanaged infra?

The [change](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4917/files#r1810976840) wasn't intentionally added, but revisiting the docs with your findings I am a little bit confused what would be the expected behavior for unmanaged on CAPA. @richardcase @nrb thoughts?

Just sharing some thoughts how we are using: we, on OpenShift, consider tags are managed by the user when they opt to provide their own network/subnets, and we don't touch on it when installing a cluster. So this change won't affect our automation as we are not setting the subnet tags for unmanaged VPC.

Reply:
For unmanaged, I'd say:

- Only apply required k8s tags to the subnets in AWS
- No other custom tags to be added to the subnets in AWS

As for syncing existing tags on the subnets back to AWSCluster, I'm not strongly opinionated on this.

Every time i look at the subnets code, it makes me 😢 IMHO, it's historically one of the most error-prone and horrible bits of code in CAPA.

Reply:
`getSubnetTagParams` still does the right thing for unmanaged VPCs (= don't use additional tags unless `TagUnmanagedNetworkResources()`), so I think the behavior is fine even with this PR. If tags are directly defined on `AWSCluster.spec.network.subnets.tags`, it's a conscious decision by the admin and those tags should be added even for unmanaged ones – why else have a field for tags?

The field description could be more precise, though (out of scope for the PR):

```
$ kubectl explain awscluster.spec.network.subnets.tags
GROUP:      infrastructure.cluster.x-k8s.io
KIND:       AWSCluster
VERSION:    v1beta2

FIELD: tags <map[string]string>


DESCRIPTION:
    Tags is a collection of tags describing the resource.
```